gateway-service-impl/src/main/java/org/hypertrace/gateway/service/logevent/LogEventsService.java

```diff
@@ -0,0 +1,172 @@
+    this.queryServiceClient = queryServiceClient;
+    this.requestTimeout = requestTimeout;
+    this.attributeMetadataProvider = attributeMetadataProvider;
+    initMetrics();
+  }
+
+  private void initMetrics() {
+    queryExecutionTimer =
+        PlatformMetricsRegistry.registerTimer(
+            "hypertrace.log.event.query.execution", ImmutableMap.of());
+  }
+
+  public LogEventsResponse getLogEventsByFilter(RequestContext context, LogEventsRequest request) {
+    Instant start = Instant.now();
+    try {
+      Map<String, AttributeMetadata> attributeMap =
+          attributeMetadataProvider.getAttributesMetadata(context, LOG_EVENT_SCOPE);
+      LogEventsResponse.Builder logEventResponseBuilder = LogEventsResponse.newBuilder();
+
+      List<LogEvent> logEvents = fetchLogEvents(context, request, attributeMap);
+
+      logEventResponseBuilder.addAllLogEvents(logEvents);
+
+      LogEventsResponse response = logEventResponseBuilder.build();
+      LOG.debug("Log Event Service Response: {}", response);
+      return response;
+    } finally {
+      queryExecutionTimer.record(
+          Duration.between(start, Instant.now()).toMillis(), TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private List<LogEvent> fetchLogEvents(
+      RequestContext context,
+      LogEventsRequest request,
+      Map<String, AttributeMetadata> attributeMetadataMap) {
+
+    AttributeMetadata timestampAttributeMetadata =
+        getTimestampAttributeMetadata(attributeMetadataProvider, context, LOG_EVENT_SCOPE);
+    QueryRequest.Builder queryBuilder =
+        QueryRequest.newBuilder()
+            .setFilter(
+                QueryAndGatewayDtoConverter.addTimeAndSpaceFiltersAndConvertToQueryFilter(
+                    convertFromMillis(
+                        request.getStartTimeMillis(), timestampAttributeMetadata.getUnit()),
+                    convertFromMillis(
+                        request.getEndTimeMillis(), timestampAttributeMetadata.getUnit()),
+                    "",
+                    timestampAttributeMetadata.getId(),
+                    "",
+                    request.getFilter()));
+
+    if (!request.getSelectionList().isEmpty()) {
+      request
+          .getSelectionList()
+          .forEach(
+              exp ->
+                  queryBuilder.addSelection(
+                      QueryAndGatewayDtoConverter.convertToQueryExpression(exp)));
+    }
+
+    addSortLimitAndOffset(request, queryBuilder);
+
+    List<LogEvent> logEventResult = new ArrayList<>();
+    QueryRequest queryRequest = queryBuilder.build();
+
+    Iterator<ResultSetChunk> resultSetChunkIterator =
+        queryServiceClient.executeQuery(queryRequest, context.getHeaders(), requestTimeout);
+
+    ResultSetMetadata resultSetMetadata = null;
+    while (resultSetChunkIterator.hasNext()) {
+      ResultSetChunk chunk = resultSetChunkIterator.next();
+
+      LOG.debug("Received chunk: {}", chunk);
+
+      if (chunk.getRowCount() < 1) {
```

Comment:
I'm guessing this is from elsewhere, but I wonder what the point of this condition is? Even if we get a chunk with 0 rows, the code handles it correctly.

Reply:
without this condition `logEventResult.add(logEventBuilder.build());` would be executed

Reply:
Isn't that inside the for loop which would never run? `for (Row row : chunk.getRowList()) {`